alot/crypto.py

```diff
@@ -81,69 +55,93 @@ def get_key(keyid, validate=False, encrypt=False, sign=False,
+            # over it.
+            try:
+                for k in keys:
+                    try:
+                        validate_key(k, encrypt=encrypt, sign=sign)
+                    except GPGProblem:
+                        # if the key is invalid for given action skip it
+                        continue
+
+                    if valid_key:
+                        # we have already found one valid key and now we find
+                        # another? We really received an ambiguous keyid
+                        raise GPGProblem(
+                            "More than one key found matching this filter. "
+                            "Please be more specific "
+                            "(use a key ID like 4AC8EE1D).",
+                            code=GPGCode.AMBIGUOUS_NAME)
+                    valid_key = k
+            except gpg.errors.GPGMEError as e:
+                # This if will be triggered if there is no key matching at all.
+                if e.getcode() == gpg.errors.AMBIGUOUS_NAME:
                     raise GPGProblem(
-                        "More than one key found matching this filter. Please "
-                        "be more specific (use a key ID like 4AC8EE1D).",
-                        code=GPGCode.AMBIGUOUS_NAME)
-                valid_key = k
+                        'Can not find any key for "{}".'.format(keyid),
+                        code=GPGCode.NOT_FOUND)
+                raise
 
             if not valid_key:
                 # there were multiple keys found but none of them are valid for
                 # given action (we don't have private key, they are expired
-                # etc)
+                # etc), or there was no key at all
                 raise GPGProblem(
                     'Can not find usable key for "{}".'.format(keyid),
                     code=GPGCode.NOT_FOUND)
             return valid_key
-        elif e.code == gpgme.ERR_INV_VALUE or e.code == gpgme.ERR_EOF:
-            raise GPGProblem('Can not find key for "{}".'.format(keyid),
+        elif e.getcode() == gpg.errors.INV_VALUE:
+            raise GPGProblem('Cannot find key for "{}".'.format(keyid),
                              code=GPGCode.NOT_FOUND)
         else:
             raise e
     if signed_only and not check_uid_validity(key, keyid):
-        raise GPGProblem('Can not find a trusworthy key for "{}".'.format(keyid),
+        raise GPGProblem('Cannot find a trusworthy key for "{}".'.format(keyid),
                          code=GPGCode.NOT_FOUND)
     return key
 
 
 def list_keys(hint=None, private=False):
     """
-    Returns a iterator of all keys containing the fingerprint, or all keys if
+    Returns a generator of all keys containing the fingerprint, or all keys if
     hint is None.
 
+    The generator may raise exceptions of :class:gpg.errors.GPGMEError, and it
+    is the caller's responsibility to handle them.
+
```

Comment:
Yea generators are strange in that respect especially if they are not used with the default pattern of 
```py
for x in gen():
    pass
```
but
```py
g = gen()
for x in g:
    pass
```
We can also leave it like it is.